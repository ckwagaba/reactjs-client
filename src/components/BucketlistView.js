import React, { Component } from 'react';
import { connect } from 'react-redux';
import Nav from './Nav';
import Header from './Header';
import Main from './Main';
import Footer from './Footer';
import Bucketlist from './Bucketlist';
import { BASEURL } from '../Config.js';
import store from '../store/Store';

class BucketlistView extends Component {
  // initial component render
  componentDidMount() {
    // get as required
    this.getBucketlists();
  }

  // when component re-renders
  componentDidUpdate(prevProps) {
    // if the current page changes, or the search term changes.
    if(prevProps.bucketlistState.currentPage !== this.props.bucketlistState.currentPage || prevProps.bucketlistState.searchTerm !== this.props.bucketlistState.searchTerm) {
      this.getBucketlists();
    }
  }

  // update currentPage
  setCurrentPage = (currentPage) => {
    store.dispatch({
      type: 'SET_CURRENT_PAGE',
      payload: currentPage
    });
  }

  // get bucketlists
  getBucketlists = () => {
    this.getTotalItems();
    fetch(BASEURL + '/bucketlists/?q=' + this.props.bucketlistState.searchTerm + '&limit=' + this.props.bucketlistState.limit + '&page=' + this.props.bucketlistState.currentPage, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': localStorage.getItem('ACCESSTOKEN')
      }
    })
    .then(response => {
      return response.json()
    })
    .then(responseData => {
      store.dispatch({
        type: 'STORE_BUCKETLISTS',
        payload: {
          listToRender: responseData.bucketlists_on_page,
          totalPages: Math.ceil(this.props.bucketlistState.totalItems / this.props.bucketlistState.limit)
        }
      });
    });
  }

  // get total items
  getTotalItems = () => {
    fetch(BASEURL + '/bucketlists/?q=' + this.props.bucketlistState.searchTerm, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': localStorage.getItem('ACCESSTOKEN')
      }
    })
    .then(response => {
      return response.json()
    })
    .then(responseData => {
      store.dispatch({
        type: 'SET_TOTAL_BUCKETLISTS',
        payload: responseData.number_of_bucketlists_on_page
      });
    });
  }

  // search API for items
  handleSearch = (event) => {
    store.dispatch({
      type: 'SET_SEARCH_TERM',
      payload: {
        searchTerm: event.target.value,
        currentPage: 1 //this is a hack. needs a solution
      }
    });
  }

  render () {
    // render bucketlists
    const rows = this.props.bucketlistState.listToRender.map((item) =>
      <Bucketlist key={item.id} bucketlistName={item.name} bucketlistId={item.id} getBucketlists={this.getBucketlists} />
    );

    // create pageButtons
    const pageButtons = [];
    for (let i = 0; i < this.props.bucketlistState.totalPages; i++) {
      pageButtons.push(<span className="page_button" key={i+1} onClick={() => this.setCurrentPage(i+1)}>{i+1}</span>);
    }

    return (
      <div className="landing_page bucketlist_view">
        <Nav handleLogout={this.props.handleLogout} />
        <Header currentLocation="Bucketlists" itemForm='/bucketlistform' searchTerm={this.props.bucketlistState.searchTerm} handleSearch={this.handleSearch} />
        <Main componentToRender={rows} />
        <Footer pageButtons={pageButtons} />
      </div>
    );
  }
}

const mapStateToProps = (store) => {
  return {
    bucketlistState: store.bucketlistView
  };
}

export default connect(mapStateToProps)(BucketlistView);
