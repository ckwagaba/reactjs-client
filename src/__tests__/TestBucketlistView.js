import React from 'react';
import { shallow } from 'enzyme';
import store from '../store/Store';
import BucketlistView from '../components/BucketlistView';

describe('BucketlistView component', () => {

  const component = shallow(
    <BucketlistView store={store} />
  );

  it('should exist', () => {
    expect(component).toMatchSnapshot();
  });

});
