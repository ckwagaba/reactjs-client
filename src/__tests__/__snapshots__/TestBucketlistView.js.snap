// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BucketlistView component should exist 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BucketlistView />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <withRouter(Nav)
          handleLogout={undefined}
/>,
        <Header
          currentLocation="Bucketlists"
          handleSearch={[Function]}
          itemForm="/bucketlistform"
          searchTerm=""
/>,
        <Main
          componentToRender={Array []}
/>,
        <Footer
          pageButtons={
                    Array [
                              <span
                                className="page_button"
                                onClick={[Function]}
                    >
                                1
                    </span>,
                            ]
          }
/>,
      ],
      "className": "landing_page bucketlist_view",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleLogout": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentLocation": "Bucketlists",
          "handleSearch": [Function],
          "itemForm": "/bucketlistform",
          "searchTerm": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "componentToRender": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "pageButtons": Array [
            <span
              className="page_button"
              onClick={[Function]}
>
              1
</span>,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <withRouter(Nav)
            handleLogout={undefined}
/>,
          <Header
            currentLocation="Bucketlists"
            handleSearch={[Function]}
            itemForm="/bucketlistform"
            searchTerm=""
/>,
          <Main
            componentToRender={Array []}
/>,
          <Footer
            pageButtons={
                        Array [
                                    <span
                                      className="page_button"
                                      onClick={[Function]}
                        >
                                      1
                        </span>,
                                  ]
            }
/>,
        ],
        "className": "landing_page bucketlist_view",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleLogout": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentLocation": "Bucketlists",
            "handleSearch": [Function],
            "itemForm": "/bucketlistform",
            "searchTerm": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "componentToRender": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "pageButtons": Array [
              <span
                className="page_button"
                onClick={[Function]}
>
                1
</span>,
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
